src/Features/CSharp/Portable/UseRecursivePatterns/CSharpUseRecursivePatternsCodeRefactoringProvider.AnalyzedNode.cs

```diff
@@ -0,0 +1,319 @@
+            }
+
+            protected static AnalyzedNode Visit(PatternMatch left, VarPattern right)
+            {
+                if (left.Expression is IdentifierNameSyntax identifier)
+                    if (SyntaxFactory.AreEquivalent(right.Identifier, identifier.Identifier))
+                        return new Conjuction(right, left.Pattern);
+
+                return null;
+            }
+
+            protected static AnalyzedNode Visit(Conjuction left, ConstantPattern right) { throw new NotImplementedException(); }
+            protected static AnalyzedNode Visit(Conjuction left, TypePattern right) { throw new NotImplementedException(); }
+            protected static AnalyzedNode Visit(Conjuction left, NotNullPattern right) { throw new NotImplementedException(); }
+            protected static AnalyzedNode Visit(Conjuction left, VarPattern right) { throw new NotImplementedException(); }
+
+            // Conjuctive combinations
+            protected static AnalyzedNode Visit(NotNullPattern left, VarPattern right) => new Conjuction(left, right);
+            protected static AnalyzedNode Visit(PatternMatch left, TypePattern right) => new Conjuction(left, right);
+            protected static AnalyzedNode Visit(TypePattern left, VarPattern right) => new Conjuction(right, left);
+
+            // Superssesive combinations
+            protected static AnalyzedNode Visit(NotNullPattern left, NotNullPattern right) => right;
+            protected static AnalyzedNode Visit(NotNullPattern left, PatternMatch right) => right;
+            protected static AnalyzedNode Visit(NotNullPattern left, TypePattern right) => right;
+
+            // Unsupported combinations
+            protected static AnalyzedNode Visit(ConstantPattern left, ConstantPattern right) => null;
+            protected static AnalyzedNode Visit(NotNullPattern left, ConstantPattern right) => null;
+            protected static AnalyzedNode Visit(PatternMatch left, ConstantPattern right) => null;
+            protected static AnalyzedNode Visit(ConstantPattern left, TypePattern right) => null;
+            protected static AnalyzedNode Visit(ConstantPattern left, VarPattern right) => null;
+            protected static AnalyzedNode Visit(TypePattern left, TypePattern right) => null;
+            protected static AnalyzedNode Visit(VarPattern left, VarPattern right) => null;
+
+            public sealed override AnalyzedNode VisitSourcePattern(SourcePattern node) => throw ExceptionUtilities.Unreachable;
+
+            private sealed class Finder : Visitor<bool>
+            {
+                private readonly IdentifierNameSyntax _expression;
+
+                private Finder(IdentifierNameSyntax expression)
+                    => _expression = expression;
+
+                public static bool Find(ExpressionSyntax expression, AnalyzedNode node)
+                    => expression is IdentifierNameSyntax id && new Finder(id).Visit(node);
```

Comment:
> new Finder

This object will be instantiated in every unification attempt. I wonder if it worth it to add a `argument` argument to `Visit` methods so this (and `Rewriter`) can use a singleton.

Reply:
Probably not (if this is a refactoring) :)

Reply:
won't refactorings run on every line that cursor 
is placed (to show lightbulb or populate action menu)?

Reply:
this is how it looks like: https://github.com/alrz/roslyn/commit/b96563c64f039475dc3aa124c64448f1f4619330

Reply:
> won't refactorings run on every line that cursor

Yes.  And that's a very small amount of work.  Contrast that with, say, an analyzer which ends up running over thousands or millions of nodes.   This is many orders of magnitude less.  To the point that it's not something we have to be terribly concerned about by default.

Reply:
I must say I have never understood the bar for performance improvements, the fact that it doesn't run that much (but not too rarely either) is enough reason to not care? I mean, shouldn't we try to do our best when possible? Considering that this addition would add zero complexity (vs. a few kilobytes of added code size) but it would eliminate a lot of allocations in the the local run. (read this as just a general question, not an argument)